File: app/ios/Runner/video_call/VideoCallState.swift
// VideoCallState.swift
// converting from dart/flutter to swift

import Foundation
import WebRTC
import flutter_callkit_incoming

enum VideoCallStatus {
    case idle
    case calling
    case ringing
    case connected
    case disconnected
}

class VideoCallState: NSObject, ObservableObject {
    static var shared = VideoCallState()
    var connection: RTCPeerConnection?
    var showingLocalVideo = false
    var showingRemoteVideo = false
    var status = VideoCallStatus.idle
    var incomingCalls = [Any]()
    var currentVideoChat: ActiveVideoChat?
    var bottom = 5
    var right = 5
    var sizeFactor = 180
    var isConfirmingClose = false
    var isConfirmingTimer: Timer? = nil
    var starting = false
    var ringTimer: Timer?
    var remoteDescriptionSet = false
    var connectionFactory: RTCPeerConnectionFactory?
    var videoCapturer: RTCCameraVideoCapturer?
    var peerConnectionFactory = RTCPeerConnectionFactory()
    var localVideoSource: RTCVideoSource?
    var localVideoTrack: RTCVideoTrack?
    var iceCandidateBuffer: [[String: Any]] = []
    var connectionOpen = false
    var localAudioSource: RTCAudioSource?
    var localAudioTrack: RTCAudioTrack?
    var currentCallData: flutter_callkit_incoming.Data?

    @Published var remoteStream: RTCMediaStream?

    func initializePeerConnection() {
        let configuration = RTCConfiguration()
        configuration.iceServers = [RTCIceServer(urlStrings: ["stun:stun.l.google.com:19302"])]
        connectionFactory = RTCPeerConnectionFactory()
        let connection = connectionFactory?.peerConnection(with: configuration, constraints: RTCMediaConstraints(mandatoryConstraints: nil, optionalConstraints: nil), delegate: MyPeerConnectionDelegate.shared)
        self.connection = connection!
    }

    func endVideoChatReceived(data: [String: [String: Any]]) {
        let callRoom = data["callRoom"]!
        let callRoomId = callRoom["id"] as! Int
        if currentVideoChat?.callRoom?.id == callRoomId {
            closeVideoCall()
        }
    }

    func closeVideoCall() {
        if connection == nil {
            return
        }
        DispatchQueue.main.async {
            let appDelegate = UIApplication.shared.delegate as! AppDelegate
            appDelegate.closeWebRTCView()
            let con = self.connection
            self.connection = nil
            con?.close()
            SwiftFlutterCallkitIncomingPlugin.sharedInstance?.endCall(self.currentCallData!)
            if self.currentVideoChat != nil {
                callMethod(name: "postEndVideoChat", args: [
                    ["callRoomId": self.currentVideoChat!.callRoom?.id]
                ], completion: { result in
                    print("result of end video chat \(result)")
                })
                self.currentVideoChat = nil
            }
            
            self.localAudioTrack?.isEnabled = false
            self.localVideoTrack?.isEnabled = false
            if let remoteVideoTrack = self.remoteStream?.videoTracks.first {
                remoteVideoTrack.isEnabled = false
            }

            self.showingLocalVideo = false
            self.showingRemoteVideo = false
            self.connection = nil
            self.remoteStream = nil
            self.iceCandidateBuffer = []
            self.remoteDescriptionSet = false
            self.isConfirmingClose = false
            self.iceCandidateBuffer = []
            self.connectionOpen = false
            self.localAudioSource = nil
            self.localAudioTrack = nil
            self.localVideoTrack = nil
            self.localVideoSource = nil
            self.currentCallData = nil
        }
    }
    
    func videoChatCallRequest(eventResponse: [String: Any]) {
        let data = eventResponse["data"] as! [String: Any]
        print("video chat call request")
        incomingCalls.append(data)
    }
    
    func videoChatSdpOffer(eventResponse: [String: Any]) {
        ringTimer?.invalidate()
        ringTimer = nil
        let data = eventResponse["data"] as! [String: Any]
        let sdp = data["sdp"] as! [String: Any]
        setSdp(sdp: sdp)
    }
    
    func setSdp(sdp: [String: Any]) {
        let remoteDescription = RTCSessionDescription(type: RTCSdpType.toType(typeString: sdp["type"] as! String)!, sdp: sdp["sdp"] as! String)
        connection!.setRemoteDescription(remoteDescription, completionHandler: { [self] error in
            if let error = error {
                print("Failed to set remote description: \(error)")
                return
            }
            print("set remote description successfully")
//            self.connection?.answer(for: RTCMediaConstraints(mandatoryConstraints: nil, optionalConstraints: nil), completionHandler: { answer, error in
//                if let error = error {
//                    print("Failed to create answer: \(error)")
//                    return
//                }
//                guard let answer = answer else { return }
//                self.connection?.setLocalDescription(answer, completionHandler: { error in
//                    if let error = error {
//                        print("Failed to set local description: \(error)")
//                        return
//                    }
//                    
//                    callMethod(name: "postCallRoomAnswer", args: [["sdp": answer.toMap()], currentVideoChat?.callRoom?.id], completion: { result in
//                        print("post call room answer result", result)
//                    })
//                })
//            })
        })
    }

    func videoChatSdpAnswer(eventResponse: [String: Any]) {
        let data = eventResponse["data"] as! [String: Any]
        iceCandidateBuffer = []
        let sdp = data["sdp"] as! [String: Any]
        let remoteDescription = RTCSessionDescription(type: RTCSdpType.toType(typeString: sdp["type"] as! String)!, sdp: sdp["sdp"] as! String)
        connection?.setRemoteDescription(remoteDescription, completionHandler: { error in
            if let error = error {
                print("Failed to set remote description: \(error)")
                return
            }
            self.remoteDescriptionSet = true
            if !self.iceCandidateBuffer.isEmpty {
                for notification in self.iceCandidateBuffer {
                    self.iceCandidate(eventResponse: notification)
                }
            }
        })
    }

    func videoChatCallAnswer(eventResponse: [String: Any]) {
        print("got video chat call answer")
    }

    func iceCandidate(eventResponse: [String: Any]) {
        print("got ice candidate")
         let data = eventResponse["data"] as! [String: Any]
        let candidateInfo = data["candidate"] as! [String: Any]
        let candidate = RTCIceCandidate(sdp: candidateInfo["candidate"] as! String, sdpMLineIndex: candidateInfo["sdpMLineIndex"] as! Int32, sdpMid: candidateInfo["sdpMid"] as? String)
        if connection != nil {
            print("adding ice candidate")
            connection?.add(candidate)
        } else {
            print("error adding ice candidate because connection is nil")
        }
    }

    func videoChatReject(call: [String: Any]) {
        var callRoom = call["callRoom"] as! [String: Any]
        callMethod(name: "postEndVideoChat", args: [
            ["callRoomId": callRoom["id"]]
        ], completion: { result in
            print("result of end video chat \(result)")
        })
        // incomingCalls.removeAll(where: { $0 as! [String: Any] == call })
        // remove by call room id
        incomingCalls.removeAll(where: { ($0 as! [String: [String: Any]])["callRoom"]?["id"] as! Int == callRoom["id"] as! Int })
    }

    func getLocalStream() {
        self.localVideoSource = peerConnectionFactory.videoSource()
        self.videoCapturer = RTCCameraVideoCapturer(delegate: self.localVideoSource!)
        self.localVideoTrack = peerConnectionFactory.videoTrack(with: self.localVideoSource!, trackId: "localVideoTrack")
        
        self.localAudioSource = peerConnectionFactory.audioSource(with: RTCMediaConstraints(mandatoryConstraints: nil, optionalConstraints: nil))
        self.localAudioTrack = peerConnectionFactory.audioTrack(with: localAudioSource!, trackId: "localAudioTrack")

        guard
            let frontCamera = (RTCCameraVideoCapturer.captureDevices().first { $0.position == .front }),
        
            // choose highest res
            let format = (RTCCameraVideoCapturer.supportedFormats(for: frontCamera).sorted { (f1, f2) -> Bool in
                let width1 = CMVideoFormatDescriptionGetDimensions(f1.formatDescription).width
                let width2 = CMVideoFormatDescriptionGetDimensions(f2.formatDescription).width
                return width1 < width2
            }).last,
        
            // choose highest fps
            let fps = (format.videoSupportedFrameRateRanges.sorted { return $0.maxFrameRate < $1.maxFrameRate }.last) else {
            return
        }
        
        self.videoCapturer!.startCapture(with: frontCamera, format: format, fps: Int(fps.maxFrameRate))
        
        showingLocalVideo = true
    }

    // to receive a call, you initiate a WebRTC session
    func receiveCall(incomingCall: [String: Any]) {
        print("receiving call")
        // incomingCalls.removeAll(where: { $0 as! [String: Any] == incomingCall })
        let callRoomData = incomingCall["callRoom"] as! [String: Any]
        incomingCalls.removeAll(where: { ($0 as! [String: [String: Any]])["callRoom"]?["id"] as! Int == callRoomData["id"] as! Int })
        // incomingCall["users"] is [{id: 1}, {id: 2}]
        let usersData = incomingCall["users"] as! [[String: Any]]
        // get list of ids as int
        let recipients = usersData.map({ user in user["id"] as! Int })
        
        currentVideoChat = ActiveVideoChat.createFromMap(json: [
            "callRoom": incomingCall["callRoom"]!,
            "recipients": recipients,
            "originator": incomingCall["originator"]!,
            "active": true
        ])
        iceCandidateBuffer = []
        
        showingLocalVideo = true
        initializePeerConnection()
        getLocalStream()
        sendSdpOffer()
    }
    
    func sendSdpOffer(renegotiation: Bool = false) {
        let offer = self.connection?.offer(for: RTCMediaConstraints(mandatoryConstraints: nil, optionalConstraints: nil), completionHandler: { offer, error in
            if let error = error {
                print("Failed to create offer: \(error)")
                return
            }
            guard let offer = offer else { return }
            self.connection?.setLocalDescription(offer, completionHandler: { error in
                if let error = error {
                    print("Failed to set local description: \(error)")
                    return
                }
                callMethod(name: "postVideoChatSdpOffer", args: [
                    ["callRoomId": self.currentVideoChat!.callRoom?.id, "authenticatedUserIds": self.currentVideoChat!.recipients, "sdp": offer.toMap(), "renegotiation": renegotiation]
                ], completion: { result in
                    if let result = result.data as? [String: Any] {
                        if self.currentVideoChat == nil {
                            self.currentVideoChat = ActiveVideoChat.createFromMap(json: result)
                        } else {
                            self.currentVideoChat?.callRoom = CallRoom.createFromMap(json: result["callRoom"] as! [String : Any])
                            self.currentVideoChat?.recipients = result["users"] as! [Int]
                        }
                        print("entered video call room and did postVideoChatSdpOffer", result)
                        self.connectionOpen = true
                        if self.iceCandidateBuffer.count > 0 {
                            for candidate in self.iceCandidateBuffer {
                                self.sendIceCandidate(candidateInfo: candidate)
                            }
                            self.iceCandidateBuffer = []
                        }
                        if renegotiation == false {
                            print("adding local video track")
                            self.connection!.add(self.localVideoTrack!, streamIds: ["localVideoStream"])
                            self.connection!.add(self.localAudioTrack!, streamIds: ["localAudioStream"])

                        }
                    }
                })
            })
        })
    }

    func startCall(recipients: [Int]) {
        if starting {
            return
        }
        starting = true
        getLocalStream()
        callMethod(name: "postManageCallRoom", args: [
            ["callRoomId": nil, "authenticatedUserIds": recipients, "sdp": nil]
        ], completion: { result in
            let data = result.data as! [String: Any?]
            if self.currentVideoChat == nil {
                self.currentVideoChat = ActiveVideoChat.createFromMap(json: data["users"] as! [String: Any])
            } else {
                self.currentVideoChat?.callRoom = CallRoom.createFromMap(json: data["callRoom"] as! [String: Any])
                self.currentVideoChat?.recipients = data["users"] as! [Int]
            }
            self.ringTimer = Timer.scheduledTimer(withTimeInterval: 25, repeats: false, block: { _ in
                self.closeVideoCall()
            })
        })
        starting = false
    }
    
    func sendIceCandidate(candidateInfo: [String: Any]) {
        callMethod(name: "postIceCandidate", args: [[
            "candidate": candidateInfo,
            "callRoomId":  VideoCallState.shared.currentVideoChat?.callRoom?.id! as Any,
            "sdp": VideoCallState.shared.connection?.localDescription?.toMap() as Any
        ]]) { result in
            print("post ice candidate result", result)
        }
    }
}

File: app/lib/views/video_view/video_call_state.dart
import 'dart:async';

import 'package:dev/lib/call_method.dart';
import 'package:dev/lib/show_alert.dart';
import 'package:dev/views/video_view/components/active_video_chat.dart';
import 'package:dev/views/video_view/components/incoming_call.dart';
import 'package:dev/views/video_view/video_call.dart';
import 'package:flutter_callkit_incoming/flutter_callkit_incoming.dart';
import 'package:flutter_webrtc/flutter_webrtc.dart';
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'package:uuid/uuid.dart';

enum VideoCallStatus {
  idle,
  calling,
  ringing,
  connected,
  disconnected,
}

ClientVideoChatData? currentVideoCall;

class ClientVideoChatData {
  var remoteVideoPromises = [];
  RTCPeerConnection? connection;
  RTCSessionDescription? remoteDescription;
  var incomingIceCandidates = [];
  var localSdpId = 0;
  var localAppliedSdpId = 0;
  var remoteSdpId = 0;
  // MediaStream? localStream;
  var localStreamAttachedToConnection = false;
  RTCSessionDescription? currentOffer;
  RTCSessionDescription? currentAnswer;
  String? uuid;
  CallRoom? callRoom;
  List<dynamic> bufferedIceCandidates = [];

  constructor() {
    uuid = Uuid().v4();
  }

  sendSdpOffer() async {
    currentOffer = await connection?.createOffer();
    localSdpId += 1;

    var result = callMethod("postVideoChatSdpOffer", [
      {"callRoomId": callRoom?.id, "offer": currentOffer?.toMap(), "sdpId": localSdpId}
    ]);
    return result;
  }

  sendSdpAnswer() async {
    localSdpId += 1;
    var answer = await currentVideoCall?.connection?.createAnswer();
    await currentVideoCall?.connection?.setLocalDescription(answer!);
    var result = await callMethod("postVideoChatSdpAnswer", [
      {"answer": answer?.toMap(), "callRoomId": callRoom?.id, "sdpId": localSdpId, "remoteSdpId": remoteSdpId}
    ]);
    if (showHttpErrorIfExists(result)) {
      return;
    }
  }

  trySetRemoteDescription(MediaStream localStream, sdp, targetSdpId, remoteSdpId) async {
    print("trying to set remote description for sdpId $targetSdpId remoteSdpId $remoteSdpId");
    print("current sdp id ${currentVideoCall?.localSdpId}, current remot sdp ${currentVideoCall?.remoteSdpId}");
    if (localSdpId == targetSdpId && currentVideoCall!.remoteSdpId < remoteSdpId) {
      if (currentVideoCall?.currentOffer != null) {
        print(">>>>>>> setting current offer");
        currentVideoCall?.connection?.setLocalDescription(currentVideoCall!.currentOffer!);
        currentVideoCall?.currentOffer = null;
      }
      currentVideoCall?.remoteSdpId = remoteSdpId;
      currentVideoCall?.remoteDescription = RTCSessionDescription(sdp["sdp"], sdp["type"]);
      print("setting remote description");
      currentVideoCall?.connection?.setRemoteDescription(currentVideoCall!.remoteDescription!).catchError((error) => {
            print("error setting remote description in videoChatSdpAnswer"),
            showAlert("Error setting remote description in videoChatSdpAnswer: $error")
          });
      if (currentVideoCall?.localStreamAttachedToConnection == false) {
        print("ATTACHING LOCAL STREAM");
        currentVideoCall?.attachLocalStreamToConnection(localStream);
      }
      if (currentVideoCall?.bufferedIceCandidates.isNotEmpty == true) {
        for (var i = 0; i < (currentVideoCall?.bufferedIceCandidates.length ?? 0); ++i) {
          var data = currentVideoCall?.bufferedIceCandidates[i];
          currentVideoCall?.bufferedIceCandidates.removeAt(i);
          i--;
          print("trying to add buffered candidate");
          tryAddIceCandidate(data);
        }
      }
    }
  }

  tryAddIceCandidate(data) async {
    if (currentVideoCall?.localSdpId == data.remoteSdpId) {
      print('will add cancdidate');
      var candidate = RTCIceCandidate(
        data['candidate']['candidate'],
        data['candidate']['sdpMid'],
        data['candidate']['sdpMLineIndex'],
      );
      currentVideoCall?.connection?.addCandidate(candidate).catchError((error) => {print("Error adding ice candidate")});
    } else {
      print("will skip candidate");
    }
  }

  attachLocalStreamToConnection(localStream) {
    localStreamAttachedToConnection = true;
    var tracks = localStream?.getTracks();
    for (var i = 0; i < (tracks?.length ?? 0); i++) {
      print("Adding track $i ${tracks[i]}");
      connection?.addTrack(tracks![i], localStream!);
    }
  }

  close() {
    remoteDescription = null;
    incomingIceCandidates = [];
    localStreamAttachedToConnection = false;
    localSdpId = 0;
    remoteSdpId = 0;
    currentAnswer = null;
    currentOffer = null;
    bufferedIceCandidates = [];
    if (connection != null) {
      connection?.onAddTrack = null;
      // @ts-ignore
      connection?.onRemoveTrack = null;
      // @ts-ignore
      connection?.onRemoveStream = null;
      connection?.onIceCandidate = null;
      connection?.onIceConnectionState = null;
      connection?.onSignalingState = null;
      connection?.onIceGatheringState = null;
      connection?.onRenegotiationNeeded = null;
      connection?.close();
      connection = null;
    }
  }
}

class VideoCallState extends ChangeNotifier {
  // RTCPeerConnection? connection;
  MediaStream? localStream;
  MediaStream? remoteStream;
  MediaStream? remoteAudioStream;
  RTCVideoRenderer? localRenderer;
  RTCVideoRenderer? remoteRenderer;
  RTCVideoRenderer? remoteAudioRenderer;
  bool showingLocalVideo = false;
  bool showingRemoteVideo = false;
  VideoCallStatus status = VideoCallStatus.idle;
  List<dynamic> incomingCalls = [];
  ActiveVideoChat? currentVideoChat;
  double bottom = 5;
  double right = 5;
  double sizeFactor = 180;
  bool isConfirmingClose = false;
  Timer? isConfirmingTimer = null;
  bool starting = false;
  Timer? ringTimer;
  bool localTracksAdded = false;
  String? currentCallkitId;

  update() {
    notifyListeners();
  }

  get width {
    return (localRenderer?.value.aspectRatio ?? 0) * sizeFactor;
  }

  get height {
    return 1 / (localRenderer?.value.aspectRatio ?? 1) * sizeFactor;
  }

  initializePeerConnection(ClientVideoChatData clientVideoChatData) async {
    var connection = await createPeerConnection({
      'iceServers': [
        {'urls': 'stun:stun.l.google.com:19302'},
        {'urls': "stun:stun4.l.google.com:5349"},
      ]
    });

    connection.onIceCandidate = (RTCIceCandidate candidate) async {
      print("ICE CANDIDATE");
      // print("got ice candidate $candidate");
      callMethod('postIceCandidate', [
        {
          "callRoomId": clientVideoChatData.callRoom?.id,
          "candidate": candidate.toMap(),
          "sdp": (await connection.getLocalDescription())?.toMap(),
          "sdpId": clientVideoChatData.localSdpId,
          "remoteSdpId": clientVideoChatData.remoteSdpId
        }
      ]).onError((error, stackTrace) => showAlert("Error sending ice candidate"));
    };

    connection.onRenegotiationNeeded = () async {
      print("renegotiation needed");
      try {
        await currentVideoCall!.sendSdpOffer();
      } catch (err, stackTrace) {
        print("Error in on renegotiation needed");
        print(err);
        print(stackTrace);
      }
    };

    connection.onIceConnectionState = (RTCIceConnectionState state) {
      if (state == RTCIceConnectionState.RTCIceConnectionStateFailed || state == RTCIceConnectionState.RTCIceConnectionStateClosed) {
        closeVideoCall();
      }
    };
    connection.onIceGatheringState = (RTCIceGatheringState state) {
      print("ICE gathering state: $state");
      if (state == RTCIceGatheringState.RTCIceGatheringStateComplete) {
        print("ICE gathering complete");
      }
    };
    connection.onSignalingState = (RTCSignalingState state) {
      print("Signaling state: $state");
    };

    connection.onTrack = (RTCTrackEvent event) async {
      print("will be adding track $event");
      // for (var stream in  event.streams) {
      if (event.track.kind == "video") {
        remoteStream = event.streams[0];
        // remoteRenderer!.setSrcObject(stream: event.streams[0], trackId: "remote" + event.track.kind.toString())
        remoteRenderer!.srcObject = remoteStream;
        showingRemoteVideo = true;
      } else if (event.track.kind == "audio") {
        remoteAudioRenderer!.srcObject = event.streams[0];
        remoteAudioStream = event.streams[0];
      }
      // }
      notifyListeners();
    };
    print("connection initialized");
    clientVideoChatData.connection = connection;
    return connection;
  }

  endVideoChatReceived(Map data) {
    var callRoomId = data['data']?['callRoom']?['id'];
    print("call room id is $callRoomId");
    if (currentVideoChat?.callRoom?.id == callRoomId) {
      closeVideoCall();
    }
  }

  startCall(List recipients) async {
    print("will try to connect");
    if (starting) return;
    try {
      starting = true;
      currentVideoCall?.close();
      currentVideoCall = ClientVideoChatData();
      localStream = await getLocalStream();
      initPlayers();
      var result = await callMethod('postCreateCallRoom', [
        {
          "callRoomId": null,
          "authenticatedUserIds": recipients,
        }
      ]);
      print("result is $result");
      if (showHttpErrorIfExists(result)) {
        closeVideoCall();
        return;
      }
      currentVideoChat = ActiveVideoChat.createFromMap({
        'callRoom': result['callRoom'],
        'recipients': result['users'].map((user) => user['id']),
        'originator': result['originator'],
        'active': false,
      });
      // currentVideoChat = ActiveVideoChat.createFromMap({'recipients': result['users'], 'callRoom': result['callRoom']});
      currentVideoCall?.callRoom = currentVideoChat?.callRoom;
      notifyListeners();
      ringTimer = Timer(const Duration(seconds: 25), () {
        closeVideoCall();
      });
    } catch (err, stackTrace) {
      print("Error starting video call");
      print(err);
      print(stackTrace);
      showAlert("Error starting video call");
      return;
    } finally {
      starting = false;
    }
  }

  void closeVideoCall({String? callkitId}) async {
    if (currentVideoCall == null) {
      return;
    }
    var call = currentVideoCall;
    currentVideoCall = null;
    call?.close();

    showingLocalVideo = false;
    showingRemoteVideo = false;
    localStream?.dispose();
    localStream = null;
    remoteStream?.dispose();
    remoteStream = null;
    remoteAudioStream?.dispose();
    remoteAudioStream = null;
    localRenderer?.dispose();
    localRenderer = null;
    remoteRenderer?.dispose();
    remoteRenderer = null;
    remoteAudioRenderer?.dispose();
    remoteAudioRenderer = null;
    iceCandidateNotifications = [];
    remoteDescriptionSet = false;
    isConfirmingClose = false;
    localTracksAdded = false;
    if (callkitId != null || currentCallkitId != null) {
      FlutterCallkitIncoming.endCall(currentCallkitId ?? callkitId ?? "").catchError((error) => print("error"));
    }
    if (currentVideoChat != null) {
      var result = await callMethod('postEndVideoChat', [
        {"callRoomId": currentVideoChat!.callRoom?.id}
      ]);
      print("result of end video chat $result");
      currentVideoChat = null;
    }
    notifyListeners();
  }

  rejectCall(incomingCall) async {
    await callMethod("postEndVideoChat", [
      {"callRoomId": incomingCall.callRoom.id}
    ]);
    print("incoming call $incomingCall");
    incomingCalls.remove(incomingCall);
  }

  void videoChatCallRequest(Map<String, dynamic> eventResponse) {
    var data = eventResponse['data'];
    print("video chat call request");
    incomingCalls.add(data);
    notifyListeners();
  }

  void videoChatSdpOffer(Map<String, dynamic> eventResponse) async {
    print("got video chat sdp offer");
    // return;
    ringTimer?.cancel();
    ringTimer = null;
    var data = eventResponse['data'];
    print("received video chat sdp offer");
    if (currentVideoCall == null) {
      print("not doing anything with sdp offer since current video cal is null");
      return;
    }
    if (currentVideoCall!.remoteSdpId < currentVideoCall!.localSdpId) {
      print("not doing negotiation because haven't received answer yet");
      return;
    }
    if (currentVideoCall?.connection == null) {
      print("must make connection receiving sdp offer");
      await initializePeerConnection(currentVideoCall!);
    }
    if (data["sdpId"] > currentVideoCall!.remoteSdpId) {
      currentVideoCall!.remoteSdpId = data["sdpId"];
      currentVideoCall!.remoteDescription = RTCSessionDescription(data["offer"]["sdp"], data["offer"]["type"]);
      currentVideoCall!.connection!.setRemoteDescription(currentVideoCall!.remoteDescription!);
      currentVideoCall!.sendSdpAnswer();
    }
    if (currentVideoCall?.localStreamAttachedToConnection == false) {
      currentVideoCall!.attachLocalStreamToConnection(localStream);
    }

    await initPlayers();
  }

  var remoteDescriptionSet = false;
  List<Map<String, dynamic>> iceCandidateNotifications = [];
  void videoChatSdpAnswer(Map<String, dynamic> eventResponse) async {
    print("GOT ANSWER");
    var data = eventResponse['data'];
    print("video chat sdp answer");
    currentVideoCall?.trySetRemoteDescription(localStream!, data["answer"], data["remoteSdpId"], data["sdpId"]);
  }

  void videoChatCallAnswer(Map<String, dynamic> eventResponse) async {
    print("got video chat call answer");
    throw UnimplementedError();
  }

  void iceCandidate(Map<String, dynamic> eventResponse) async {
    print("ICE CANDIDATE");
    try {
      if (!remoteDescriptionSet) {
        iceCandidateNotifications.add(eventResponse);
        return;
      }

      var data = eventResponse['data'];

      print("got ice candidate, adding to connection");
      if (currentVideoCall == null || currentVideoCall?.callRoom?.id != data["callRoom"]["id"]) {
        currentVideoCall?.close();
        currentVideoCall = ClientVideoChatData();
        await initializePeerConnection(currentVideoCall!);
      }
      if (currentVideoCall!.remoteSdpId < data["sdpId"]) {
        currentVideoCall!.bufferedIceCandidates.add(data);
        return;
      }
      if (!currentVideoCall!.localStreamAttachedToConnection) {
        currentVideoCall!.attachLocalStreamToConnection(localStream);
      }
      currentVideoCall!.tryAddIceCandidate(data);

      // wait one second
      await Future.delayed(const Duration(seconds: 1));
      print("Done waiting!");
    } catch (err, stackTrace) {
      print("Error in ice candidate");
      print(err);
      print(stackTrace);
    }
  }

  void videoChatReject(Map call) {
    closeVideoCall();
    print("rejecting call $call");
    callMethod("postEndVideoChat", [
      {"callRoomId": call["callRoom"]["id"]}
    ]);
    incomingCalls.remove(call);
    notifyListeners();
  }

  getLocalStream() async {
    var stream = await navigator.mediaDevices.getUserMedia({
      'video': {
        'mandatory': {
          'minWidth': '640', // Provide your own width, height and frame rate here
          'minHeight': '480',
          'minFrameRate': '15',
        },
        'facingMode': 'user',
        'optional': [],
      },
      'audio': true,
    });
    showingLocalVideo = true;
    notifyListeners();
    return stream;
  }

  initPlayers() async {
    if (localRenderer == null) {
      var tempLocalRenderer = RTCVideoRenderer();
      var tempRemoteRenderer = RTCVideoRenderer();
      var tempRemoteAudioRenderer = RTCVideoRenderer();
      Helper.setSpeakerphoneOn(true);
      await Future.wait([
        tempLocalRenderer.initialize(),
        tempRemoteRenderer.initialize(),
        tempRemoteAudioRenderer.initialize(),
      ]);
      localRenderer = tempLocalRenderer;
      remoteRenderer = tempRemoteRenderer;
      remoteAudioRenderer = tempRemoteAudioRenderer;
    }
    localRenderer?.srcObject = localStream;
  }

  void receiveCall(Map incomingCall, {String? callkitId}) async {
    print("receiving call $incomingCall");
    // close keyboard if visible
    FocusManager.instance.primaryFocus?.unfocus();

    currentCallkitId = callkitId;
    currentVideoCall?.close();
    currentVideoCall = ClientVideoChatData();

    incomingCalls.remove(incomingCall);
    print("receiving call $incomingCall");

    currentVideoChat = ActiveVideoChat.createFromMap({
      'callRoom': incomingCall['callRoom'],
      'recipients': incomingCall['users'].map((user) => user['id']),
      'originator': incomingCall['originator'],
      'active': true,
    });
    currentVideoCall?.callRoom = currentVideoChat?.callRoom;
    print("current video chat $currentVideoChat");
    notifyListeners();

    await initializePeerConnection(currentVideoCall!);
    showingLocalVideo = true;
    localStream = await getLocalStream();
    await initPlayers();
    var result = await currentVideoCall?.sendSdpOffer();
    if (showHttpErrorIfExists(result)) {
      closeVideoCall();
      return;
    }
    notifyListeners();
  }
}

var videoCallState = VideoCallState();
